File: packages/modules/narduk-logging/src/logger.ts
import { createConsola, LogLevels } from 'consola/core'
import {
  boundRecord,
  cleanText,
  freezeRecord,
  sanitizeErrorForLog,
  sanitizeFields,
  sanitizeUrlForLog,
} from './sanitize.js'
import { createConsoleSink } from './sinks.js'
import type {
  ConfiguredLogLevel,
  LogContext,
  LogDiagnostics,
  LogLevel,
  Logger,
  LoggerOptions,
  LogRecord,
  LogSink,
  SafeFields,
} from './types.js'

const CONTEXT_LIMITS = { requestId: 128, operationId: 128, method: 32, path: 512 } as const

export function resolveLogLevel(
  value: unknown,
  fallback: ConfiguredLogLevel = 'info',
): ConfiguredLogLevel {
  return typeof value === 'string' &&
    ['trace', 'debug', 'info', 'warn', 'error', 'fatal', 'silent'].includes(value)
    ? (value as ConfiguredLogLevel)
    : fallback
}

function identity(value: string | undefined, name: string, limit: number): string {
  // eslint-disable-next-line no-control-regex -- Identity labels must reject control characters.
  if (
    typeof value !== 'string' ||
    !value.trim() ||
    value.length > limit ||
    /[\u0000-\u001f\u007f]/.test(value)
  ) {
    throw new TypeError(
      `Logging ${name} must be a non-empty string of at most ${limit} characters without controls`,
    )
  }
  return value.trim()
}

type SafeContext = Omit<LogContext, 'data'> & { data?: SafeFields }

function safeContext(context: LogContext = {}, options: LoggerOptions): SafeContext {
  const fields = sanitizeFields(context, options)
  const result: SafeContext = {}
  for (const [key, limit] of Object.entries(CONTEXT_LIMITS)) {
    const value = fields[key]
    if (typeof value === 'string') {
      const name = key as keyof typeof CONTEXT_LIMITS
      result[name] =
        name === 'path' ? sanitizeUrlForLog(value).slice(0, limit) : value.slice(0, limit)
    }
  }
  if (typeof fields.traceId === 'string' && /^[a-f0-9]{32}$/.test(fields.traceId))
    result.traceId = fields.traceId
  if (typeof fields.spanId === 'string' && /^[a-f0-9]{16}$/.test(fields.spanId))
    result.spanId = fields.spanId
  if (['server', 'client', 'job', 'cli'].includes(String(fields.source)))
    result.source = fields.source as LogContext['source']
  if (fields.data && typeof fields.data === 'object' && !Array.isArray(fields.data))
    result.data = fields.data
  return result
}

/** Creates an isolated logger. Only explicit sinks can start asynchronous delivery. */
export function createLogger(options: LoggerOptions): Logger {
  const base = {
    service: identity(options.service, 'service', 128),
    environment: identity(options.environment, 'environment', 64),
    runtime: identity(options.runtime ?? 'javascript', 'runtime', 64),
    ...(options.release ? { release: identity(options.release, 'release', 128) } : {}),
  }
  const level = resolveLogLevel(
    options.level,
    options.environment === 'development' ? 'debug' : 'info',
  )
  const sinks = [
    ...(options.sinks ?? [
      createConsoleSink(
        options.format ?? (options.environment === 'development' ? 'pretty' : 'json'),
      ),
    ]),
  ]
  const diagnostics: LogDiagnostics = { emitted: 0, sinkFailures: 0, dropped: 0 }
  let closed = false
  let emitting = false
  const engine = createConsola({
    level: LogLevels[level],
    throttle: 0,
    reporters: [
      {
        log(entry) {
          const record = entry.args[0] as Readonly<LogRecord>
          diagnostics.emitted++
          for (const sink of sinks) {
            try {
              sink.write(record)
            } catch {
              diagnostics.sinkFailures++
            }
          }
        },
      },
    ],
  })

  async function lifecycle(method: 'flush' | 'close'): Promise<void> {
    const results = await Promise.allSettled(
      sinks.map(async (sink: LogSink) => {
        await sink[method]?.()
      }),
    )
    diagnostics.sinkFailures += results.filter((result) => result.status === 'rejected').length
  }

  function scoped(context: SafeContext, scope?: string): Logger {
    const emit = (target: LogLevel, message: string, data?: Record<string, unknown>): void => {
      if (LogLevels[target] > engine.level) return
      if (closed || emitting) {
        diagnostics.dropped++
        return
      }
      emitting = true
      try {
        const safeData = data ? sanitizeFields(data, options) : undefined
        const { data: boundData, ...bindings } = context
        const fields = { ...boundData, ...safeData }
        const error = fields.error
        delete fields.error
        const record: LogRecord = {
          schemaVersion: 1,
          timestamp: (options.clock?.() ?? new Date()).toISOString(),
          ...base,
          ...bindings,
          level: target,
          message: cleanText(typeof message === 'string' ? message : '[Invalid message]'),
          ...(scope ? { scope } : {}),
          ...(Object.keys(fields).length ? { data: fields } : {}),
        }
        if (error !== undefined) record.error = sanitizeErrorForLog(error, options)
        engine[target]({ args: [freezeRecord(boundRecord(record))] })
      } catch {
        diagnostics.dropped++
      } finally {
        emitting = false
      }
    }
    const bind = (next: LogContext): SafeContext => {
      const safe = safeContext(next, options)
      return { ...context, ...safe, data: { ...context.data, ...safe.data } }
    }
    const logger: Logger = {
      trace: (message, data) => emit('trace', message, data),
      debug: (message, data) => emit('debug', message, data),
      info: (message, data) => emit('info', message, data),
      warn: (message, data) => emit('warn', message, data),
      error: (message, data) => emit('error', message, data),
      fatal: (message, data) => emit('fatal', message, data),
      child: (name, next = {}) =>
        scoped(bind(next), cleanText(scope ? `${scope}.${name}` : name, 256)),
      withContext: (next) => scoped(bind(next), scope),
      async operation(name, work, data) {
        const operationLog = logger.withContext({ operationId: crypto.randomUUID() })
        const start = performance.now()
        try {
          const result = await work(operationLog)
          operationLog.info('Operation completed', {
            ...data,
            operation: name,
            outcome: 'success',
            durationMs: Math.round(performance.now() - start),
          })
          return result
        } catch (error) {
          operationLog.error('Operation failed', {
            ...data,
            operation: name,
            outcome: 'failure',
            durationMs: Math.round(performance.now() - start),
            error,
          })
          throw error
        }
      },
      flush: () => lifecycle('flush'),
      async close() {
        if (!closed) {
          closed = true
          await lifecycle('close')
        }
      },
      get diagnostics() {
        return { ...diagnostics }
      },
    }
    return logger
  }
  return scoped(safeContext(options.context, options))
}

File: packages/modules/narduk-logging/python/src/narduk_logging/logger.py
import asyncio
import re
import time
import uuid
from collections.abc import Awaitable, Callable, Mapping, MutableMapping
from dataclasses import dataclass, field, replace
from datetime import UTC, datetime
from threading import Lock, local
from typing import NoReturn, TypeVar, cast

import structlog

from .sanitize import MAX_RECORD_BYTES, clean_text, sanitize_error, sanitize_fields, sanitize_url
from .sinks import StreamSink, copy_record, encode_record
from .types import LogLevel, Record, Sink

_PRIORITY = {"trace": 0, "debug": 1, "info": 2, "warn": 3, "error": 4, "fatal": 5, "silent": 6}
_CONTEXT_LIMITS = {"requestId": 128, "operationId": 128, "method": 32, "path": 512}
T = TypeVar("T")


@dataclass(frozen=True)
class LoggerOptions:
    service: str
    environment: str
    runtime: str = "python"
    release: str | None = None
    level: LogLevel | None = None
    redact: tuple[str, ...] = ()
    include_stack: bool = False
    clock: Callable[[], datetime] = lambda: datetime.now(UTC)


@dataclass
class Diagnostics:
    emitted: int = 0
    sink_failures: int = 0
    dropped: int = 0


@dataclass
class _State:
    options: LoggerOptions
    sinks: tuple[Sink, ...]
    lock: Lock = field(default_factory=Lock)
    reentry: local = field(default_factory=local)
    diagnostics: Diagnostics = field(default_factory=Diagnostics)
    closed: bool = False


def _identity(value: str, name: str, limit: int) -> str:
    if not value.strip() or len(value) > limit or re.search(r"[\x00-\x1f\x7f]", value):
        raise ValueError(f"Logging {name} must contain 1..{limit} characters without controls")
    return value.strip()


def _safe_context(fields: Mapping[str, object], options: LoggerOptions) -> Record:
    safe = sanitize_fields(dict(fields), redact=options.redact, include_stack=options.include_stack)
    result: Record = {}
    for name, limit in _CONTEXT_LIMITS.items():
        value = safe.get(name)
        if isinstance(value, str):
            result[name] = sanitize_url(value) if name == "path" else value[:limit]
    for name, length in (("traceId", 32), ("spanId", 16)):
        value = safe.get(name)
        if isinstance(value, str) and re.fullmatch(f"[a-f0-9]{{{length}}}", value):
            result[name] = value
    if safe.get("source") in ("server", "client", "job", "cli"):
        result["source"] = safe["source"]
    if isinstance(safe.get("data"), dict):
        result["data"] = safe["data"]
    return result


class Logger:
    def __init__(self, state: _State, context: Record | None = None, scope: str = "") -> None:
        self._state = state
        self._context = copy_record(context or {})
        self._scope = scope
        # wrap_logger creates an isolated processor chain without structlog.configure().
        self._engine = structlog.wrap_logger(
            structlog.ReturnLogger(),
            processors=[self._process],
            wrapper_class=structlog.BoundLogger,
        )

    @property
    def diagnostics(self) -> Diagnostics:
        with self._state.lock:
            return replace(self._state.diagnostics)

    def _process(
        self, _wrapped: object, method: str, event: MutableMapping[str, object]
    ) -> NoReturn:
        options = self._state.options
        minimum = options.level or ("debug" if options.environment == "development" else "info")
        if _PRIORITY[method] < _PRIORITY[minimum]:
            raise structlog.DropEvent
        with self._state.lock:
            if self._state.closed or getattr(self._state.reentry, "busy", False):
                self._state.diagnostics.dropped += 1
                raise structlog.DropEvent
        self._state.reentry.busy = True
        try:
            self._emit(method, event)
        except Exception:
            # Logging failures cannot replace the application's error or change its result.
            with self._state.lock:
                self._state.diagnostics.dropped += 1
        finally:
            self._state.reentry.busy = False
        raise structlog.DropEvent

    def _emit(self, level: str, event: MutableMapping[str, object]) -> None:
        options = self._state.options
        fields = sanitize_fields(
            event.get("fields", {}), redact=options.redact, include_stack=options.include_stack
        )
        bindings = copy_record(self._context)
        bound_data = bindings.pop("data", {})
        data = {**(bound_data if isinstance(bound_data, dict) else {}), **fields}
        error = data.pop("error", None)
        message = event.get("event")
        record: Record = {
            "schemaVersion": 1,
            "timestamp": options.clock()
            .astimezone(UTC)
            .isoformat(timespec="milliseconds")
            .replace("+00:00", "Z"),
            "service": options.service,
            "environment": options.environment,
            "runtime": options.runtime,
            **bindings,
            "level": level,
            "message": clean_text(message) if isinstance(message, str) else "[Invalid message]",
        }
        if options.release:
            record["release"] = options.release
        if self._scope:
            record["scope"] = self._scope
        if data:
            record["data"] = data
        if error is not None:
            record["error"] = sanitize_error(error, include_stack=options.include_stack)
        if len(encode_record(record).encode()) > MAX_RECORD_BYTES:
            record["data"] = {"truncated": True}
            if isinstance(error, dict):
                record["error"] = {
                    "name": error.get("name", "Error"),
                    "message": error.get("message", "Error"),
                }
        if len(encode_record(record).encode()) > MAX_RECORD_BYTES:
            record["message"] = clean_text(str(record["message"]), 512)
            if "error" in record:
                record["error"] = {"name": "Error", "message": "[Truncated error]"}
        with self._state.lock:
            self._state.diagnostics.emitted += 1
        for sink in self._state.sinks:
            try:
                sink.write(
                    copy_record(record)
                )  # Independent snapshots prevent cross-sink mutation.
            except Exception:
                with self._state.lock:
                    self._state.diagnostics.sink_failures += 1

    def trace(self, message: str, data: Mapping[str, object] | None = None) -> None:
        self._engine.trace(message, fields=data or {})

    def debug(self, message: str, data: Mapping[str, object] | None = None) -> None:
        self._engine.debug(message, fields=data or {})

    def info(self, message: str, data: Mapping[str, object] | None = None) -> None:
        self._engine.info(message, fields=data or {})

    def warn(self, message: str, data: Mapping[str, object] | None = None) -> None:
        self._engine.warn(message, fields=data or {})

    def error(self, message: str, data: Mapping[str, object] | None = None) -> None:
        self._engine.error(message, fields=data or {})

    def fatal(self, message: str, data: Mapping[str, object] | None = None) -> None:
        self._engine.fatal(message, fields=data or {})

    def with_context(self, **context: object) -> "Logger":
        safe = _safe_context(context, self._state.options)
        old_data = self._context.get("data", {})
        new_data = safe.get("data", {})
        return Logger(
            self._state,
            {
                **self._context,
                **safe,
                "data": {
                    **(old_data if isinstance(old_data, dict) else {}),
                    **(new_data if isinstance(new_data, dict) else {}),
                },
            },
            self._scope,
        )

    def child(self, scope: str, **context: object) -> "Logger":
        child = self.with_context(**context)
        return Logger(
            self._state,
            child._context,
            clean_text(f"{self._scope}.{scope}" if self._scope else scope, 256),
        )

    def operation(
        self, name: str, work: Callable[["Logger"], T], data: Mapping[str, object] | None = None
    ) -> T:
        child = self.with_context(operationId=str(uuid.uuid4()))
        start = time.monotonic()
        try:
            value = work(child)
        except BaseException as error:
            child.error(
                "Operation failed",
                {
                    **(data or {}),
                    "operation": name,
                    "outcome": "failure",
                    "durationMs": round((time.monotonic() - start) * 1000),
                    "error": error,
                },
            )
            raise
        child.info(
            "Operation completed",
            {
                **(data or {}),
                "operation": name,
                "outcome": "success",
                "durationMs": round((time.monotonic() - start) * 1000),
            },
        )
        return value

    async def aoperation(
        self,
        name: str,
        work: Callable[["Logger"], Awaitable[T]],
        data: Mapping[str, object] | None = None,
    ) -> T:
        child = self.with_context(operationId=str(uuid.uuid4()))
        start = time.monotonic()
        try:
            value = await work(child)
        except BaseException as error:
            child.error(
                "Operation failed",
                {
                    **(data or {}),
                    "operation": name,
                    "outcome": "failure",
                    "durationMs": round((time.monotonic() - start) * 1000),
                    "error": error,
                },
            )
            raise
        child.info(
            "Operation completed",
            {
                **(data or {}),
                "operation": name,
                "outcome": "success",
                "durationMs": round((time.monotonic() - start) * 1000),
            },
        )
        return value

    def flush(self) -> None:
        self._lifecycle("flush")

    def close(self) -> None:
        with self._state.lock:
            if self._state.closed:
                return
            self._state.closed = True
        self._lifecycle("close")

    async def aclose(self) -> None:
        await asyncio.to_thread(self.close)

    def _lifecycle(self, method: str) -> None:
        for sink in self._state.sinks:
            try:
                if method == "close":
                    sink.close()
                else:
                    sink.flush()
            except Exception:
                with self._state.lock:
                    self._state.diagnostics.sink_failures += 1


def create_logger(
    *,
    service: str,
    environment: str,
    runtime: str = "python",
    release: str | None = None,
    level: LogLevel | None = None,
    sinks: tuple[Sink, ...] | None = None,
    redact: tuple[str, ...] = (),
    include_stack: bool = False,
    clock: Callable[[], datetime] | None = None,
) -> Logger:
    options = LoggerOptions(
        service=_identity(service, "service", 128),
        environment=_identity(environment, "environment", 64),
        runtime=_identity(runtime, "runtime", 64),
        release=_identity(release, "release", 128) if release else None,
        level=cast(LogLevel | None, level if level in _PRIORITY else None),
        redact=redact,
        include_stack=include_stack,
        clock=clock or (lambda: datetime.now(UTC)),
    )
    return Logger(
        _State(
            options,
            sinks if sinks is not None else (StreamSink(pretty=environment == "development"),),
        )
    )
